### PR TITLE
[libc++] Deprecate and remove meaningless `<cxxx>` headers

### LINKED ISSUE #37759: The standard library header <cstdalign> is missing
|  |  |
| --- | --- |
| Bugzilla Link | [38411](https://llvm.org/bz38411) |
| Version | unspecified |
| OS | FreeBSD |
| CC | @mclow |

## Extended Description 
I found this problem on FreeBSD, which uses clang and llvm C++ library: https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=230230

cppreference says that there should be such standard header: https://en.cppreference.com/w/cpp/header/cstdalign

### LINKED ISSUE #99985: P0619R4: Reviewing Deprecated Facilities of C++17 for C++20
**Link:** https://wg21.link/p0619r4

```
BEGIN-RST-NOTES
Only sections D.7, D.8, D.9, D.10, D.11, D.12 and D.13 are implemented. Section D.4 remains undone.
END-RST-NOTES
```

### PR DESCRIPTION
**Pull request by frederick-vs-ja**
This PR deprecates `<ccomplex>`, `<cstdbool>`, `<ctgmath>`, and `<ciso646>` in C++17 and "removes" them in C++20 by special deprecation warnings.

`<cstdalign>` is previously missing. This PR also tries to add them, and then deprecates and "removes" `<cstdalign>`. Fixes #37759.

Papers:
- https://wg21.link/P0063R3
- https://wg21.link/P0619R4

Closes #99985.

### CODE REVIEW

File: libcxx/include/stdalign.h [hunk 1]
```diff
@@ -0,0 +1,42 @@
+// -*- C++ -*-
```
Comment (philnik777) [inline libcxx/include/stdalign.h:1]: We don't need a wrapper here, since clang provides a correct one.

File: libcxx/include/ccomplex [hunk 2]
```diff
@@ -23,4 +23,17 @@
...
+#    define _DEPRECATION_ANNOTATION _LIBCPP_DEPRECATED_("removed in C++20")
+#  else
+#    define _DEPRECATION_ANNOTATION _LIBCPP_DEPRECATED
+#  endif
+
+using __standard_header_ccomplex _DEPRECATION_ANNOTATION = void;
+using __use_standard_header_ccomplex                     = __standard_header_ccomplex;
+
+#  undef _DEPRECATION_ANNOTATION
+#endif
```
Comment (philnik777) [inline libcxx/include/ccomplex:37]: This seems like a very complex way of writing
```c++
#if defined(__DEPRECATED) && __DEPRECATED
#  warning "<ccomplex> is deprecated"
#endif
```

File: libcxx/include/ciso646 [hunk 3]
```diff
@@ -21,4 +21,9 @@
 #  pragma GCC system_header
 #endif
 
+#if _LIBCPP_STD_VER >= 20
+using __standard_header_ciso646 _LIBCPP_DEPRECATED_("removed in C++20") = void;
+using __use_standard_header_ciso646                                     = __standard_header_ciso646;
+#endif
```
Comment (philnik777) [inline libcxx/include/ciso646:27]: I'm not sure it's a good idea to deprecate `<ciso646>` without an escape hatch specifically for this, since it's used in the wild quite a bit. This will likely cause major breakage.

**Review by philnik777 — CHANGES_REQUESTED** [review]
We could also consider removing most of the headers as a DR. Most of these headers are likely almost never used in the wild except for `<ciso646>` (as our missing `<cstdalign>` shows). @ldionne @jwakely @STL any thoughts?

File: libcxx/include/ciso646 (hunk 3 above)
Comment (frederick-vs-ja) [reply libcxx/include/ciso646]: `_LIBCPP_ENABLE_REMOVED_CISO646` added.

File: libcxx/include/stdalign.h (hunk 1 above)
Comment (frederick-vs-ja) [reply libcxx/include/stdalign.h]: Seems needed to make CI pass.

Comment (philnik777) [reply libcxx/include/stdalign.h:1]: Could you explain?

**Comment by jwakely** [comment]
I don't think we'll ever physically remove these from libstdc++, certainly not as a DR. But I should add `#warning` to them saying they're deprecated/removed.

File: libcxx/include/stdalign.h (hunk 1 above)
Comment (frederick-vs-ja) [reply libcxx/include/stdalign.h:1]: When I attempted to just remove `stdaligh.h`, there was build error. Perhaps CMakelist.txt should also be modified.

I'm not sure whether we can/should just be inconsistent for `stdaligh.h`, given libc++ is providing almost all wrapper C headers required by the C++ standard library.

File: libcxx/include/ccomplex (hunk 2 above)
Comment (frederick-vs-ja) [reply libcxx/include/ccomplex:37]: We're only doing this for `<hash_map>` & `<hash_set>`. I _guess_ the deprecation warnings should be controllable by `_LIBCPP_DISABLE_DEPRECATION_WARNINGS` (which is mentioned in `__config`).

File: libcxx/include/stdalign.h (hunk 1 above)
Comment (philnik777) [reply libcxx/include/stdalign.h:1]: Given that we didn't have `stdalign.h` before there shouldn't be a build error. Also note that we only provide C headers which we have to for conformance. We don't provide `assert.h` or `limits.h` for example.

File: libcxx/include/ccomplex (hunk 2 above)
Comment (philnik777) [reply libcxx/include/ccomplex:37]: Yeah, we should probably slap a `&& !defined(_LIBCPP_DISABLE_DEPRECATION_WARNINGS)` at the end.

File: libcxx/include/cstdalign [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+    cstdalign synopsis
+
+Macros:
+
+    __alignas_is_defined
+    __alignof_is_defined
+
+*/
+
+#include <__config>
```
Comment (philnik777) [inline libcxx/include/cstdalign:23]: This should include `<stdalign.h>` in case some vendor adds any extensions to that header.

File: libcxx/docs/ReleaseNotes/20.rst [hunk 5]
```diff
@@ -38,8 +38,9 @@ What's New in Libc++ 20.0.0?
 Implemented Papers
 ------------------
 
-- P2747R2: ``constexpr`` placement new (`Github <https://github.com/llvm/llvm-project/issues/105427>`__)
+- P0619R4: Reviewing Deprecated Facilities of C++17 for C++20 (`Github <https://github.com/llvm/llvm-project/issues/99985>`__)
 - P2609R3: Relaxing Ranges Just A Smidge (`Github <https://github.com/llvm/llvm-project/issues/105253>`__)
+- P2747R2: ``constexpr`` placement new (`Github <https://github.com/llvm/llvm-project/issues/105427>`__)
```
Comment (philnik777) [inline libcxx/docs/ReleaseNotes/20.rst:43]: Unrelated changes?

File: libcxx/docs/Status/Cxx17Papers.csv [hunk 6]
```diff
@@ -43,7 +43,7 @@
 "","","","","",""
 "`P0032R3 <https://wg21.link/P0032R3>`__","Homogeneous interface for variant, any and optional","2016-06 (Oulu)","|Complete|","4.0",""
 "`P0040R3 <https://wg21.link/P0040R3>`__","Extending memory management tools","2016-06 (Oulu)","|Complete|","4.0",""
-"`P0063R3 <https://wg21.link/P0063R3>`__","C++17 should refer to C11 instead of C99","2016-06 (Oulu)","|Complete|","7.0",""
+"`P0063R3 <https://wg21.link/P0063R3>`__","C++17 should refer to C11 instead of C99","2016-06 (Oulu)","|Complete|","7.0","``<ccomplex>``, ``<cstdalign>``, ``<cstdbool>``, and ``<ctgmath>`` are deprecated since LLVM 20."
```
Comment (philnik777) [inline libcxx/docs/Status/Cxx17Papers.csv:46]: I don't think it makes a ton of sense to note deprecations here. This has nothing to do with this paper.

File: libcxx/docs/ReleaseNotes/20.rst (hunk 5 above)
Comment (frederick-vs-ja) [reply libcxx/docs/ReleaseNotes/20.rst:43]: Oh, I intentionally sorted the bullets and the sorting itself was perhaps unrelated. Perhaps I should drop the sorting from this PR.

Should we sort the bullets in the release notes at some time?

File: libcxx/docs/Status/Cxx17Papers.csv (hunk 6 above)
Comment (frederick-vs-ja) [reply libcxx/docs/Status/Cxx17Papers.csv]: Reverted.

File: libcxx/include/cstdalign (hunk 4 above)
Comment (frederick-vs-ja) [reply libcxx/include/cstdalign:23]: Added now.

File: libcxx/include/ccomplex [hunk 7]
```diff
@@ -23,4 +23,12 @@
 #  pragma GCC system_header
 #endif
 
+#if _LIBCPP_STD_VER >= 17 && defined(__DEPRECATED) && __DEPRECATED && !defined(_LIBCPP_DISABLE_DEPRECATION_WARNINGS)
```
Comment (ldionne) [inline libcxx/include/ccomplex:26]: What is the purpose of checking `__DEPRECATED` here?

File: libcxx/include/tgmath.h [hunk 8]
```diff
@@ -24,7 +24,12 @@
 #endif
 
 #ifdef __cplusplus
-#  include <ctgmath>
+#  if _LIBCPP_STD_VER >= 17
```
Comment (ldionne) [inline libcxx/include/tgmath.h:27]: Can't we always include `<cmath>` and `<complex>` instead?

File: libcxx/include/complex.h [hunk 9]
```diff
@@ -24,7 +24,11 @@
 #endif
 
 #ifdef __cplusplus
-#  include <ccomplex>
+#  if _LIBCPP_STD_VER >= 17
+#    include <complex>
+#  else
+#    include <ccomplex>
+#  endif
```
Comment (ldionne) [inline libcxx/include/complex.h:31]: Similarly, why not include `<complex>` unconditionally instead?

File: libcxx/test/std/depr/depr.cpp.headers/ciso646.verify.cpp [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+
+// <ciso646>
+
+// check that <ciso646> is removed in C++20
+// When built with modules, <ciso646> should be omitted.
+
+// UNSUPPORTED: c++03, c++11, c++14, c++17
+// UNSUPPORTED: clang-modules-build
+
+#include "test_macros.h"
```
Comment (ldionne) [inline libcxx/test/std/depr/depr.cpp.headers/ciso646.verify.cpp:17]: This include is not necessary.

File: libcxx/test/std/language.support/support.runtime/cstdalign.pass.cpp [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+
+#ifdef alignas
+#  error alignas should not be defined
+#endif
+
+#ifdef alignof
+#  error alignof should not be defined
+#endif
+
+int main(int, char**) { return 0; }
```
Comment (ldionne) [inline libcxx/test/std/language.support/support.runtime/cstdalign.pass.cpp:31]: I think this can be a `.compile.pass.cpp`?

File: libcxx/docs/ReleaseNotes/20.rst [hunk 12]
```diff
@@ -82,6 +83,11 @@ Deprecations and Removals
   were private but could cause ambiguity in name lookup. Code that expects such ambiguity will possibly not compile in
   LLVM 20.
 
+- ``<ccomplex>``, ``<cstdalign>`` (previously missing), ``<cstdbool>``, and ``<ctgmath>`` are deprecated since C++17 as
+  specified by the standard. They, together with ``<ciso646>``, are removed in C++20, but libc++ still provides these
+  headers as an extension and only deprecates them. The ``_LIBCPP_ENABLE_REMOVED_CISO646`` macro can be defined to
+  suppress deprecation for ``<ciso646>``.
```
Comment (ldionne) [inline libcxx/docs/ReleaseNotes/20.rst:89]: Why are we treating `<ciso646>` differently? At first glance my preference would be to handle it consistently with the other headers and only provide `_LIBCPP_DISABLE_DEPRECATION_WARNINGS`.

File: libcxx/include/ccomplex (hunk 7 above)
Comment (philnik777) [reply libcxx/include/ccomplex:26]: It's so people don't get deprecation warnings when specifying `-Wno-deprecated`.

File: libcxx/include/tgmath.h (hunk 8 above)
Comment (frederick-vs-ja) [reply libcxx/include/tgmath.h]: Yeah. I'm correspondingly changing `<ctgmath>` and `<ccomplex>` now.

File: libcxx/include/stdalign.h (hunk 1 above)
Comment (jwakely) [reply libcxx/include/stdalign.h]: We certainly do provide stdatomic.h

File: libcxx/include/ccomplex (hunk 7 above)
Comment (ldionne) [reply libcxx/include/ccomplex:26]: Should we instead set `_LIBCPP_DISABLE_DEPRECATION_WARNINGS` when `-Wno-deprecated` is passed?

**Comment by ldionne** [comment]
I'm fine with the patch in principle, but we do have a problem that the modules for these now-deprecated headers get built whenever the `std` module is built, which breaks the compilation of anything that builds the `std` module. I pulled this down locally to play around and even making these modules `explicit module ccomplex` doesn't fix the issue, so I'm not certain how to fix this.

@ian-twilightcoder Do you have ideas? Basically, the TLDR is that we have headers that now contain an unconditional `#warning` directive and we'd like these headers not to be pulled into the `std` module's compilation unless they are requested explicitly. Making these headers `textual` solves the problem but that seems like a hack.

**Comment by ian-twilightcoder** [comment]
> I'm fine with the patch in principle, but we do have a problem that the modules for these now-deprecated headers get built whenever the `std` module is built, which breaks the compilation of anything that builds the `std` module. I pulled this down locally to play around and even making these modules `explicit module ccomplex` doesn't fix the issue, so I'm not certain how to fix this.
> 
> @ian-twilightcoder Do you have ideas? Basically, the TLDR is that we have headers that now contain an unconditional `#warning` directive and we'd like these headers not to be pulled into the `std` module's compilation unless they are requested explicitly. Making these headers `textual` solves the problem but that seems like a hack.

You really can't use `#error` or `#warning` in modules because the headers all build with the module, and there's no way to trigger a diagnostic on import rather than during build. `textual` will cause issues with duplicate declarations, that can only really be used if 1) there aren't declarations (e.g. x-macro headers) or 2) there is only one includer.

So for now I think you have to check if modules are enabled and don't `#warning`. We're investigating syntax to mark submodules as deprecated, and also entire headers.

File: libcxx/include/stdalign.h (hunk 1 above)
Comment (philnik777) [reply libcxx/include/stdalign.h]: @jwakely Sorry, I meant `stdalign.h`. At least it looks to me like you're not providing the required macros when including it.

Comment (jwakely) [reply libcxx/include/stdalign.h]: Yeah, GCC's `stdalign.h` only defines them for C11 and C17, but no versions of C++. I have a local patch fixing it which I'll propose. I don't think anybody has ever noticed before, because nobody cares about those macros in C++ (so much so that one of them was overlooked until earlier this year and nobody noticed or cared). The `__alignof_is_defined` macro isn't actually mentioned in any published C++ standard.

File: libcxx/include/ccomplex (hunk 7 above)
Comment (philnik777) [reply libcxx/include/ccomplex:26]: I'm not sure. You can have `-Wno-deprecated -Wdeprecated-declarations`, which disables deprecation warnings but keeps `[[deprecated]]` warnings. I'm not sure it makes a ton of sense to support this weird combination either though.

File: libcxx/include/stdalign.h (hunk 1 above)
Comment (jwakely) [reply libcxx/include/stdalign.h]: Patch to fix GCC's `<stdalign.h>`: https://gcc.gnu.org/pipermail/gcc-patches/2024-October/666236.html

Thanks for pointing it out!

File: libcxx/include/ccomplex (hunk 2 above)
Comment (frederick-vs-ja) [reply libcxx/include/ccomplex]: I think that works as CI is green now.

File: libcxx/include/ccomplex [hunk 13]
```diff
@@ -23,4 +23,17 @@
...
+using __standard_header_ccomplex
+#  if _LIBCPP_STD_VER >= 20
+    _LIBCPP_DEPRECATED_("removed in C++20. Include <complex> instead.")
+#  else
+    _LIBCPP_DEPRECATED_("Include <complex> instead.")
+#  endif
+    = void;
+using __use_standard_header_ccomplex = __standard_header_ccomplex;
+
+#endif
```
Comment (ldionne) [inline libcxx/include/ccomplex:37]:
```suggestion
#  if _LIBCPP_STD_VER >= 20
using __standard_header_ccomplex _LIBCPP_DEPRECATED_("removed in C++20. Include <complex> instead.") = void;
  using __use_standard_header_ccomplex = __standard_header_ccomplex;
#  elif _LIBCPP_STD_VER >= 17
  using __standard_header_ccomplex _LIBCPP_DEPRECATED_("Include <complex> instead.") = void;
  using __use_standard_header_ccomplex = __standard_header_ccomplex;
#endif
```

Maybe 1-2 lines longer but easier to follow IMO.

The same comment applies to other headers.

File: libcxx/test/std/depr/depr.cpp.headers/ciso646.verify.cpp [hunk 14]
```diff
@@ -0,0 +1,18 @@
...
+//
+//===----------------------------------------------------------------------===//
+
+// <ciso646>
+
+// check that <ciso646> is removed in C++20
+// When built with modules, <ciso646> should be omitted.
+
+// UNSUPPORTED: c++03, c++11, c++14, c++17
+// UNSUPPORTED: clang-modules-build
```
Comment (ldionne) [inline libcxx/test/std/depr/depr.cpp.headers/ciso646.verify.cpp:15]: IIUC, these diagnostics are not emitted under Clang modules? So then the clever using _LIBCPP_DEPRECATED(...)` approach is equivalent to

```
#if !__has_feature(cxx_modules) // whatever the right way of checking that is
# warning "this header is deprecated"
#endif
```

I'm not super strongly attached, but I do think that using a plain `#warning` makes this easier to understand than the clever `using` trick.

Comment (ian-twilightcoder) [reply libcxx/test/std/depr/depr.cpp.headers/ciso646.verify.cpp:15]: `__has_feature(modules)` We want this for clang modules and cxx_modules is C++20 modules

Comment (frederick-vs-ja) [reply libcxx/test/std/depr/depr.cpp.headers/ciso646.verify.cpp:15]: I'm afraid that currently we can't reliably detect `-fmodules`, as `__has_feature(modules)` is currently `1` in C++20 mode even when `-fno-modules` is used, see #57432.

File: libcxx/include/ccomplex (hunk 13 above)
Comment (frederick-vs-ja) [reply libcxx/include/ccomplex]: Done.

**Review by ldionne — APPROVED** [review]
LGTM.

File: libcxx/test/std/depr/depr.cpp.headers/ciso646.verify.cpp (hunk 14 above)
Comment (philnik777) [reply libcxx/test/std/depr/depr.cpp.headers/ciso646.verify.cpp:15]: Why can't we use `#if __building_module(std)`?
